### DISCUSSION TITLE
Proposal: Guard statement

### DISCUSSION THREAD

**Discussion by scottdorman** [dotnet/csharplang#138] [General]
## Background
Swift has a guard statement that can be very valuable in writing code that is easy to reason about. See [Swift Guard Statement - why you should use it.](http://ericcerney.com/swift-guard-statement/)

## Problem
If we have a function `Foo` that takes an integer that must be positive (`> 0`), we can write that as:

```c#
Bar Foo_alt1(int b) {
    if (b > 0) {
        // all of our code is surrounded by checks for the condition.
        // special handling of condition is far away from the statement of what the condition is
        // extra indent, extra nesting, etc.
    } else {
        throw new IllegalArgumentException(nameof(b));
    }
}

Bar Foo_alt2(int b) {
    // 'syntactic sugar' disadvantage is that now we check for the "negative"
    // condition instead of the positive. this might be natural in cases of null checking or
    // other simple checks, but can become burdensome in complex cases of many comparisons
    // and logical operators.
    if (b <= 0) {
        throw new IllegalArgumentException(nameof(b));
    } else {
        // rest of code to handle the normal case is here.
        // 'excessive' nesting and indentation
        return foo;
    }
}

Bar Foo_alt3(int b) {
    // 'syntactic sugar' disadvantage is that now we check for the "negative"
    // condition instead of the positive. this might be natural in cases of null checking or
    // other simple checks, but can become burdensome in complex cases of many comparisons
    // and logical operators.
    if (b <= 0) {
        throw new IllegalArgumentException(nameof(b));
    }

    // rest of code to handle the normal case is here.
    // we choose not to resort to nesting, but in doing so, we lose an important compile-time check:
    // a programmer error might result in the if-block above not returning or throwing, thus
    // causing the control flow to drop off into the "normal" code
    return foo;
}
```

## Proposed Solution: guard statements
```c#
Bar Foo(int b) {
    // state that b must be a positive integer
    guard (b > 0) else {
        throw new IllegalArgumentException(nameof(b));
    }
    // compiler verifies that the guard-block throws or returns, thus
    // guaranteeing that code here will satisfy the guard condition.
    // ... do something with b without worrying about edge cases
}
```

## Benefits
The guard statement introduces two benefits over `if`-based guards:

1. Syntactic sugar of declaring the positive condition we would like to meet after the guard, rather than the negatives
2. Compile-time checking of the fact that the guard clause ends control flow by returning or throwing
3. One less indent :)

## Grammar

```antlr
guard-statement
    : 'guard' ( boolean-expression ) 'else' embedded-statement
    ;
```

## See Also
- https://github.com/dotnet/roslyn/issues/8181
- https://github.com/dotnet/roslyn/issues/119
- https://github.com/dotnet/roslyn/issues/6400
- https://github.com/dotnet/roslyn/issues/11562

**Comment by gulshan** [comment]
Method contracts #105 is trying to solve the same problem.

**Comment by Thaina** [comment]
@gulshan The purpose of guard statement is covered more than just parameter. It can be used on variable later on

**Comment by Thaina** [comment]
I would still go against adding `guard` keyword. I suggest `return if` syntax

https://github.com/dotnet/roslyn/issues/11562#issuecomment-223243473
``` C#
void Guard(string s)
{
    return if(!int.TryParse(s, out var value))
    {
        // handle or report the problem
        // auto return after this block
    }

    // use value
}
```

**Comment by gulshan** [comment]
@Thaina Isn't `return if(..)` very similar to already supported `if(..) return` or `if(..) throw`? The case for `guard` is to check with the positive condition- `guard(int.TryParse(s, out var value))`, not with the negative condition like `if return` or `return if`.

**Comment by Joe4evr** [comment]
You've already linked dotnet/roslyn#11562 which was on the table for the LDM team before they decided to make `out var` declarations leak into the enclosing scope instead of introducing the (potentially) confusing `guard` statement. The odds of them turning around on this decision is extremely unlikely because there is already a lot of code relying on the leaky behavior, and you know how much the language team holds to making sure valid code today is valid code forever.

**Comment by Thaina** [comment]
@gulshan Not quite

What `return if` do is, it explicitly show from the first line of block that the whole block will surely return for meeting that condition. It just scoped syntax to help code more obvious

I would compare it to `using` for dispose. It making scope to ensure that it will dispose for that block so you will never forget to add dispose after you use it. In the same manner, `guard` or `return if` doing the same thing, ensure that this `if` block will not go below if it met the condition

To put it simply. We may have code like this

```C#
void DoSomething()
{
    if(condition)
    {
        // 20 lines of code
    }
    else
    {
        // 40 lines of code
    }
}
```

We might try to make less indent

```C#
void DoSomething()
{
    if(condition)
    {
        // 20 lines of code

        return;
    }

    // 40 lines of code
}
```

But it will not so obvious. It will be a bug if we forget to add return in the last

```C#
void DoSomething()
{
    return if(condition) // Much more obvious
    {
        // 20 lines of code
    }

    // 40 lines of code
}
```

**Comment by gulshan** [comment]
@Thaina `return if` is still checking with negative condition, not the positive one like `guard`.
And I think, if there's 20 lines of code when the parameters are not valid, then the code should be restructured and this code is not a good candidate for using guard statement.

**Comment by Thaina** [comment]
@gulshan Indeed it opposite check

But this is just minor
If we use `guard` we might write `guard(obj != null)`
If we use `return if` we might write `return if(obj == null)`

Not difference. I just don't like adding new keyword while we can reuse keyword we have. `return if` not break existing code anyway

**Comment by Thaina** [comment]
@gulshan And it not only about "the parameters are not valid" there could be any kind of logic with condition that if it meet some condition it will do one branch of logic and return

```C#
void DoSomething()
{
    return if(condition0)
    {
        // 8 lines of code
    }

    return if(condition1)
    {
        // 8 lines of code
    }

    return if(condition2)
    {
        // 8 lines of code
    }

    // 12 lines of code for not met any condition above
}
```

**Comment by michaelcheers** [comment]
I don't see how
```csharp
guard (b > 0) else throw new IllegalArgumentException(nameof(b));
```
is any better than
```csharp
if (b <= 0) throw new IllegalArgumentException(nameof(b));
```
I suppose it could be easier to read and an automatic negation but I don't think that's a big enough reason to add a feature to C#.

**Reply by soraphis** [reply]
Same here. "guard ... else" is imho the same as 'if .... else': 

    if (b > 0) {} else throw new IllegalArgumentException(nameof(b));

**Comment by Thaina** [comment]
@michaelcfanning Having keyword like `guard` have benefit to constraint error in compile time

As I was state above. This feature could compare to `using` that it really done nothing. You could dispose object manually

But `using` make a block to ensure that you will not forget to dispose. And also make it explicitly safe that outside this block cannot use that variable, because it already disposed

Same reason when you have a code block that branching off. To ensure that when condition don't match it will not touch a block of code below if block. guard statement could protect you

**Comment by iam3yal** [comment]
`match` is going to be an expression so depends on the design the following might be possible in the future:

```
void DoSomething()
{
    return match(...)
    {
        // ...
    }
}
```

**Comment by lachbaer** [comment]
It looks to me as if `guard ` would be only syntactic sugar
```C#
guard (b > 0) else throw new IllegalArgumentException(nameof(b));
```
instead of
```C#
if (!( b > 0 )) throw new IllegalArgumentException(nameof(b));
```
And I don't like `return if` statements, because it partly obfuscates where and maybe whether in a middle of a function you're gonna `return`.

And just by looking at it I don't really get what
```C#
    return if(condition0)
    {
        // 8 lines of code
    }
```
will mean. `return` if condition is met, or is not met? When will the block be executed. Will it be executed when the condition is met and after that the function will be left?

As long as it is not kind of intuitive, these constructs should not be allowed!

**Comment by iam3yal** [comment]
I'm not sure why it needs to be `guard (b > 0) else { }` instead of just `guard (b > 0) { }`.

**Comment by lachbaer** [comment]
@eyalsk 
Because otherwise it would read like "guard if (b > 0) **and if it is so** process the block", instead of "guard if (b>0) **but if not** then execute the block".

**Comment by lachbaer** [comment]
`guard` could be more useful if it **enforces** an `else`.
```C#
guard ( b > 0 )
{
  ProcessB(b);
}
else throw new IllegalArgumentException(nameof(b));
```
It would be kind of equal to `if (b>0) ... else` but in the case of `guard` the `else` clause is really *enforced* by the compiler and would issue an error otherwise.
The positive block could be ommited, resulting in `guard (b>0) else { ... };`
>    // compiler verifies that the guard-block throws or returns, thus
    // guaranteeing that code here will satisfy the guard condition.

Great idea! I would expand it to applying to the `else` block of the `guard` statement only.

**Comment by iam3yal** [comment]
@lachbaer Okay, this make sense, I'd actually go with something like `unless` if it should imply the opposite of `if`.

```
unless (b > 0) {
    throw new IllegalArgumentException(nameof(b));
}
```

**Comment by orthoxerox** [comment]
IIRC, the idea behind the original guard statement was to leak variables into the outer scope and ensure that something bottom-typed like `return` or `throw` or `FailFast` is evaluated otherwise. Now that regular `if`s leak variables into the outer scope I don't see how `Assert<ArgumentNullException>(b <= 0, $"{nameof(b)} must be a positive integer");` is any worse than adding `guard`.

**Comment by HaloFour** [comment]
Without the explicit scope leakage benefits there is little reason to implement `guard` in C#.  Sure, the enforcement of exiting the current block is useful, but I don't believe that it's worth the price of admission.

@orthoxerox 

I will say that the `Assert` helper that you've demonstrates is worse in one significant way; it evaluates that expression to build the exception message whether or not the test failed.  If you had concatenated the message in your example that wouldn't be an issue as the compiler would be smart enough to optimize that away, but you used String interpolation which isn't the lightest-weight of invocations.  I'd recommend deferring that evaluation via `Func<string>` or similar.

**Comment by MovGP0** [comment]
[Code Contracts](https://msdn.microsoft.com/en-us/library/dd264808.aspx) (#105) seems to be the better idea. No need to reinvent the squared wheel.

**Comment by iam3yal** [comment]
@MovGP0 I really hope [this](https://github.com/dotnet/csharplang/issues/105) will get implemented at some point. :)

**Comment by Thaina** [comment]
@MovGP0 I think it's not the point of this feature. Code Contract is not compile time

**Comment by iam3yal** [comment]
@Thaina If #105 will get implemented then it will be part of the language and may or may not be done at compile-time depends on the design.

**Comment by MovGP0** [comment]
@Thaina To my knowledge, Code Contracts are enforced by the CLR since v4.0 (ca. 2008) at runtime. Violations cause an exception in the calling code.

**Comment by Thaina** [comment]
@MovGP0 I think so that's why it not related to guard syntax. guard syntax main purpose is to block usage of variable at compile time

**Comment by iam3yal** [comment]
@Thaina But it might get implemented as part of the language... and then the guard construct would be useless.

**Comment by gulshan** [comment]
Note: Method contracts will probably not cover pattern matching.

**Comment by Pzixel** [comment]
Code contracts should cover this.

**Comment by OmidID** [comment]
Guard has another idea. You will lock the local variable not an inveted if !!!!

```csharp
public MyViewModel ViewModel { get; set; }

public async Task DoSomethingAsync()
{
   if (ViewModel == null) return;

   // Start Long Task
   await Task.Delay(5000);

   // ViewModel possible get null here
   await ViewModel.DoSomethingElseAsync();
}

...

public void Dispose()
{
   ViewModel = null;
}

```

How guard will protect your variable here:

```csharp
public MyViewModel ViewModel { get; set; }

public async Task DoSomethingAsync()
{
   guard (var vm = ViewModel) else {
      return;
   }

   // Start Long Task
   await Task.Delay(5000);

   // vm is not null for sure
   await vm.DoSomethingElseAsync();
}
```

**Comment by Flutterish** [comment]
`unless`? I have trouble reading the `guard` syntax, it sounds weird.

**Comment by snow-jallen** [comment]
The way elixir handles this is oh so slick, and C# could do something quite similar without new keywords or anything too confusing, you just pattern match off of the function header.

```
public int DoThing(int x) when x < 2
{
  Console.WriteLine("I do this when x is small");
}

public int DoThing(int x) when x < 10
{
  Console.WriteLine("I do this when x is medium");
}

public int DoThing(int x)
{
  Console.WriteLine("I do this when x is big");
}
```

Sooo much cleaner than statements within the method - it makes it blatantly obvious when each method is called and you can guarantee that within that method x will be validated on whatever rules are necessary.

C# 9 gets us kinda sorta close, but it's still nothing near as slick as the above example. 

````
        public static int HandleNum(int x) => x switch
        {
            < 2 => handleSmallNum(x),
            < 10 => handleMediumNum(x),
            _ => handleBigNum(x)
        };

        private static int handleSmallNum(int x)
        {
            return x;
        }

        private static int handleMediumNum(int x)
        {
            return x;
        }

        private static int handleBigNum(int x)
        {
            return x;
        }
````

**Reply by HaloFour** [reply]
See: https://github.com/dotnet/csharplang/discussions/4182

TL;DR:  Unlike functional languages, in C# overload resolution is handled at compile-time, not at runtime.  Failure to resolve would end up being a runtime exception, not a compiler error.  The lexical ordering of the member declarations is also not important in C#, but it becomes critically important in this model due to having to establish a priority for the patterns.

**Reply by danfma** [reply]
That's a good point but IMHO I don't see the possibility of generating runtime-only errors as a problem.
It will be just syntactic sugar for common error-handling scenarios.

```c#
public int DoThing(int x) when x < 2
{
  Console.WriteLine("I do this when x is small");
}
```

Above, could be translated to:

```c#
public int DoThing(int x)
{
  Contract.Requires(x < 2, nameof(x), "No satisfies 'x < 2' constraint");
  Console.WriteLine("I do this when x is small");
}
```

However, I know this is an old topic because it is just the contracts history coming back again.

**Comment by marchy** [comment]
### 'Guard Let' Consideration

This proposal and discussion missed out on what is one of the _most fundamental_ and valuable uses of Swift's `guard` statement, and that is the `guard let` part.

In addition to inverting the 'if' check, the `if let` / `guard let` keywords ALSO check for nullability, and in guard's case allows code below the statement to not need to take on another level of nesting if the condition fails. This mix of declaration pattern and forced code block exit (ie: `return`/`continue`) give it unique abilities that you CANNOT practically achieve without it, it is the realm of compilers to generate such code consistently and with the right error semantics always in place.

**Given:**
```
record SomeObject (string? someProperty);
```

**Currently (without pattern-matching):**
```
if( someObject.SomeProperty is null )
	return; // or 'continue'
string someProperty = someObject.SomeProperty;

// use 'someProperty', guaranteed to not be null
...
... 
...
```

**Currently (with pattern-matching):**
```
if( someObject.SomeProperty is string null ){
	// use 'someProperty', guaranteed to not be null
	...
	...
	...
} else {
	return; // or 'continue'
}

```


**With 'guard let':**
```
guard (someObject.SomeProperty is string someProperty ) else return; // or 'continue';

// use 'someProperty', guaranteed to not be null
...
...
...
```

**With current pattern

While we could hack around this with the current [Null Checks pattern matching](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/functional/pattern-matching#null-checks), the main point is that the inverse [declaration pattern](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#declaration-and-type-patterns) is not currently supported and does not allow for the same semantics to be achieved.

**Reply by svick** [reply]
[You can use the `not` pattern for this:](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEBMBGAsAKFwYgDsBXAG1IEMAjUiAAgkOtt13ToGFcBvXO/uigAsdALIAKAMoB7ALYQOlAM5K6SuRADyVAFYQAxsACUfAbxwDLdeHXHr52vYYB0M+QAUo0gA7RgCa1VCaWBBDAAGNQ1PHz8EEwsrfnMkpJQAdgBuOgB6HLppKDoAcn1pQmAYEghi0ySAX1ZEpLy6YiV6YvsIGN8of2KAGjoAc2IKKAoKiAgAEzpgaTpg0Kp6EnI6qxQMAE47aK8+/yNMrcacC9woA0L5twVlJXEd8IB+OgfeuNOgA=)

```c#
if (someObject.SomeProperty is not string someProperty)
{
    return; // or 'continue'
}

// use 'someProperty', guaranteed to not be null
```

**Reply by marchy** [reply]
@svick  You just made my day!

I've submitted [a request to update the documentation](https://github.com/dotnet/docs/issues/26855) to showcase this extremely useful scenario.

Modern, null-safe, idiomatic C# for the win!!  🎉

**Reply by OFark** [reply]
or `return if(someObject.SomeProperty is not string someProperty)`
